text/trace/0168-sampling-propagation.md

```diff
@@ -0,0 +1,391 @@
+
+## Explanation
+
+Two pieces of information are needed to convey consistent head trace
+sampling probability:
+
+1. The head trace sampling probability.
+2. Source of consistent sampling decisions.
+
+This proposal uses 6 bits of information for each of these and does
+not depend on built-in TraceID randomness, which is not sufficiently
+specified for probability sampling at this time.  This proposal closely 
+follows [research by Otmar Ertl](https://arxiv.org/pdf/2107.07703.pdf).
+
+### Probability value
+
+To limit the cost of this extension and for statistical reasons
+documented below, we propose to limit head trace sampling probability
+to powers of two.  This limits the available head trace sampling
+probabilities to 1/2, 1/4, 1/8, and so on.  We can compactly encode
+these probabilities as small integer values using the base-2 logarithm
+of the adjusted count.
+
+Using six bits of information we can convey known and unknown sampling
+rates as small as 2**-61.  The value 63 is reserved to mean sampling
+with probability 0, which conveys an adjusted count of 0 for the
+associated context.
+
+When propagated, the probability value will be interpreted as shown in
+the folowing table, which uses an offset of +1 in order to place the
+Unknown value at 0:
+
+| Probability Value | Head Probability | Note                   |
+| -----             | -----------      | ----                   |
+| 0                 | Unknown          | Reserved for span data |
+| 1                 | 1                |                        |
+| 2                 | 1/2              |                        |
+| 3                 | 1/4              |                        |
+| ...               | ...              |                        |
+| N                 | 2**(-N+1)        | 1 in 2**(N-1)          |
+| ...               | ...              |                        |
+| 61                | 2**-60           |                        |
+| 62                | 2**-61           |                        |
+| 63                | 0                | Maximum encoded value  |
+
+[Described in OTEP
+170](https://github.com/open-telemetry/oteps/pull/170), Span data
+sampled by the `ParentBased` sampler will encode the value that was
+propagated by the parent span as its "probability value" `p`.
+
+The value `p=0` SHOULD NOT be propagated using `tracestate`
+explicitly, because the equivalent interpretation can be obtained by
+omitting `p`.
+
+### Randomness value
+
+With head trace sampling probabilities limited to powers of two, the
+amount of randomness needed per trace context is limited.  A
+consistent sampling decision is accomplished by propagating a specific
+random variable denoted `r`.  The random variable is described by a
+(truncated) geometric distribution having shape parameter `1/2`, listed below:
+
+| `r` Value        | Selection Probability | Sampling probability   |
```

Comment:
I don't know how many people will be able to follow this. How about starting with a simpler explanation? Like generating a uniformly distributed 61-bit(ish) random value R such that a classic probabilistic sampler with probability P would sample if R <= P. That's the baseline that probably everyone will understand. Then we take into account that if P is discrete as 2**-s, then R can be similarly discreet, and its encoding can be reduced to counting the number of leading 0 bits (a couple concrete examples would probably help a lot).

The fact that r is a geometric distribution - is that relevant? Sure, we can mention it, after the more common sense explanation.

Finally, why do we need to list THAT algorithm (nextRandomness)? That's not how it will be implemented in practice in the SDKs. Why not show the more realistic algorithm that generates an n-bit random number and does bit-shifting to produce r-value?

Reply:
Thank you. I've updated this text with a more gentle introduction and removed some of the mathy words. I've changed the example code to use `rand.Int63()` and `bits.LeadingZeros64()` from the Golang standard library, also posted some code to compute it several other ways and compare the costs:

https://gist.github.com/jmacd/79c38c1056035c52f6fff7b7fc071274

Assuming an SDK uses at least 61 bits of *independent* randomness in the TraceID, it should be easy to compute in any case. The reason for this exposition is that not all TraceIDs will have independent randomness, and in that case all they need to compute an r-value, on average, are two random bits.